### DISCUSSION TITLE
anyone working on cosmopolitan ffmepg?

### DISCUSSION THREAD

**Discussion by zackees** [jart/cosmopolitan#352] [Ideas]
Would be nice to have a universal binary for ffmpeg.

**Comment by km19809** [comment]
I tried some ridiculous approach.
First, I built a `ffmpeg` WASI binary following [this repo](https://github.com/SebastiaanYN/FFmpeg-WASI).
Then, I used the [`Hermit`](https://github.com/dylibso/hermit) to convert the WASM binary into an executable.

Here's a `Hermitfile` I used for conversion:
```
FROM ffmpeg. wasm
MAP ["."]
ENV_PWD_IS_HOST_CWD
```

I didn't try the binary much, and it is slow.
However, it works, anyway.

**Reply by G4Vi** [reply]
@km19809  Cool use of hermit! Maybe I should get on adding JIT support to Hermit to make use cases like this more viable!

**Reply by jart** [reply]
I agree. The fact that Hermit is able to do this is very impressive.

**Reply by jart** [reply]
@G4Vi Getting JIT to work across platforms is tricky, even with Cosmo. Whatever JIT WASM has will most likely need to be modified to work. For example, you'll need to make sure it uses `MAP_JIT`, `__jit_begin()`, and `__jit_end()` which are the APIs that Cosmo provides which let you use RWX memory on Apple Silicon. With OpenBSD it gets a little trickier since RWX memory is forbidden entirely. You also need to use `__builtin___clear_cache(start, end)` after modifying code which is supported by Cosmo.

**Comment by jart** [comment]
I'm able to make an `ffmpeg` binary with cosmocc if I do this:

```
./configure --enable-gpl --prefix=/opt/cosmos/x86_64 --enable-cross-compile --cc=x86_64-unknown-cosmo-cc --cxx=x86_64-unknown-cosmo-c++ --ld=x86_64-unknown-cosmo-c++ --ar=x86_64-unknown-cosmo-ar --as=x86_64-unknown-cosmo-as --nm="x86_64-unknown-cosmo-nm -g"
make -j32
make install
/opt/cosmos/x86_64/bin/ffmpeg
```

It worked like a charm. Creating a truly useful ffmpeg binary would take a lot more work though. We'd need to also build/install upstream libraries like libx264, libx265, etc. For example, Debian users would expect all of these upstream libraries to be available:

```
--enable-gnutls --enable-ladspa --enable-libaom --enable-libass
--enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio
--enable-libcodec2 --enable-libdav1d --enable-libflite
--enable-libfontconfig --enable-libfreetype --enable-libfribidi
--enable-libglslang --enable-libgme --enable-libgsm --enable-libjack
--enable-libmp3lame --enable-libmysofa --enable-libopenjpeg
--enable-libopenmpt --enable-libopus --enable-libpulse
--enable-librabbitmq --enable-librist --enable-librubberband
--enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex
--enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtheora
--enable-libtwolame --enable-libvidstab --enable-libvorbis
--enable-libvpx --enable-libwebp --enable-libx265 --enable-libxml2
--enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi
--enable-lv2 --enable-omx --enable-openal --enable-opencl
--enable-opengl --enable-sdl2 --disable-sndio --enable-libjxl
--enable-pocketsphinx --enable-librsvg --enable-libmfx
--enable-libdc1394 --enable-libdrm --enable-libiec61883
--enable-chromaprint --enable-frei0r --enable-libx264
--enable-libplacebo --enable-librav1e
```

cc: @ahgamut

**Comment by syrusakbary** [comment]
Great work @jart!

For those interested in a fully working version of ffmpeg via WebAssembly, we have published one here:
https://wasmer.io/wasmer/ffmpeg

It's a bit slower than native, but it effectively uses multiple threads for encoding/decoding :)
